### ISSUE TITLE
JS/React Audit

### ISSUE THREAD

**Issue by cmc333333** [18F/omb-eregs#504]
As a modern frontend developer, I want to ensure the project is following best practices around React, building, and testings.

## AC
* Explore the existing codebase, looking for violations of best practice
* Write tests where big holes exist
* Refactor code to be more testable and in line with best practices

Our code base isn't huge, but it's also not tiny, so we don't expect this to be 100% complete. We'd like examples of the right way to do things, though.

**Comment by jeremiak** [comment]
In addition to increasing test coverage, I came up with an initial list of things though as I go through the tasks I'll make notes of opportunities for further refactoring.

Renaming/restructuring
* Ensure that there is only one component in each file to make it easier to know how components work together and where to find them. Ideally, the file name should be similar to the component name
* Remove `-view` from some filenames for consistency (`policies-view.js`, `policy-view.js`) and its mostly redundant
* Move the *Tab components out of the /pages ([`<PoliciesTab />`](https://github.com/18F/omb-eregs/blob/master/ui/pages/requirements.js#L14-L38) in requirements, for example) and into the component directory structure
* Create a /util directory to hold helper utilities so that the top level of /ui is all configuration and its easier to understand where to start
* Create an API utility, which exports { endpoints, queries } making it clearer how these things are related to each other
* Investigate removing *Container from the names of the react components in /pages. In React, Containers are frequently higher order components but below the concept of a page

New components
* Refactor the *Tab components (from the /pages components)) to be a more generic, shared component if possible. How do the current *Tab components differ from `tab-view.js`?
* A `<Link />` component that will wrap around `<ExternalLink />` and `{ Link }` from the `routes.js` - can fix https://github.com/18F/omb-eregs/issues/505 along the way
* I found `pages/search-redirect.js` a bit hard to follow, and the actual component is named `<LookupSearch />` which I confused with `lookup-search.js`. I'd like to make this component more explicit and clearer how this component is related to the other search components like `lookup-search.js` and `/components/search`

If you'd like, I can just create an issue for each one of these as I tackle them to more clearly document the scope of each piece of work.

**Comment by cmc333333** [comment]
These all sound like moves in the right direction @jeremiak, thank you! Based on [preliminary designs](https://gsa.invisionapp.com/share/QUDR1NBCK#/screens), it looks like we'll be rewriting significant chunks of the requirements page and replacing the policies page with a policy-profile. I wonder if a good use of your time would be to create a more ideal structure for those components to set a proper example for us to follow? In the meantime, we can start picking off elements in this list.

**Comment by carodew** [comment]
There will definitely be some modifications to the designs in the first prototype that you linked to – but yes, that is generally the direction. People were confused by the search results (they're still expecting a search to pull up full policy documents, not paragraphs (requirements) within documents, so there will be tweaks to that. Also we'll probably have to move in a direction of displaying the full policy text because of this strong mental model for full policy documents, but that doesn't mean we lose the requirement concept, or don't move in the direction of treating the content as more fluid and dynamic.

**Comment by cmc333333** [comment]
We got lots of useful feedback from @jeremiak and @msecret. Closing now.
